exonum_build/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+//! This crate simplifies writing build.rs for exonum and exonum services.
+
+extern crate protoc_rust;
+extern crate walkdir;
+
+use protoc_rust::Customize;
+use walkdir::WalkDir;
+
+use std::{
+    env,
+    fs::File,
+    io::Write,
+    path::{Path, PathBuf},
+};
+
+/// Finds all .proto files in `path` and subfolders and returns a vector of their paths.
+fn get_proto_files<P: AsRef<Path>>(path: &P) -> Vec<PathBuf> {
+    WalkDir::new(path)
+        .into_iter()
+        .filter_map(|e| {
+            let e = e.ok()?;
+            if e.path().extension()?.to_str() == Some("proto") {
+                Some(e.path().into())
+            } else {
+                None
+            }
+        }).collect()
+}
+
+/// Workaround for https://github.com/stepancheg/rust-protobuf/issues/324
+/// It is impossible to `include!` .rs files generated by rust-protobuf
+/// so we generate piece of `mod.rs` which includes generated files as public submodules.
+///
+/// tests.proto .rs file will be included with `#[cfg(test)]`.
+fn generate_mod_rs<P: AsRef<Path>, Q: AsRef<Path>>(
+    out_dir: &P,
+    proto_files: &[PathBuf],
+    mod_file: &Q,
+) {
+    let mod_file_content = {
+        proto_files
+            .iter()
+            .map(|f| {
+                let mod_name = f
+                    .file_stem()
+                    .unwrap()
+                    .to_str()
+                    .expect(".proto file name is not convertible to &str");
+                if mod_name == "tests" {
+                    format!("#[cfg(test)]\npub mod {};\n", mod_name)
+                } else {
+                    format!("pub mod {};\n", mod_name)
+                }
+            }).collect::<String>()
+    };
+    let dest_path = out_dir.as_ref().join(mod_file);
+    let mut file = File::create(dest_path).expect("Unable to create output file");
+    file.write_all(mod_file_content.as_bytes())
+        .expect("Unable to write data to file");
+}
+
+/// Generates .rs files from .proto files.
+///
+/// `protoc` executable from protobuf should be in `$PATH`
+///
+/// # Examples
+///
+/// In `build.rs`
+/// ```no_run
+/// extern crate exonum_build;
+///
+/// use exonum_build::protobuf_generate;
+///
+/// // Includes usually should contain input_dir.
+/// protobuf_generate("src/proto", &["src/proto"], "example_mod.rs")
+/// ```
+/// After successful run `$OUT_DIR` will contain \*.rs for each \*.proto file in
+/// "src/proto/\*\*/" and example_mod.rs which will include all generated .rs files
+/// as submodules.
+///
+/// To use generated protobuf structs.
+///
+/// In `src/proto/mod.rs`
+/// ```ignore
+/// extern crate exonum;
+///
+/// include!(concat!(env!("OUT_DIR"), "/example_mod.rs"));
+///
+/// // If you use types from exonum .proto files.
+/// use exonum::encoding::protobuf::*;
+/// ```
+pub fn protobuf_generate<P, Q, R, I>(input_dir: P, includes: I, mod_file_name: Q)
+where
+    P: AsRef<Path>,
+    Q: AsRef<Path>,
+    R: AsRef<Path>,
+    I: IntoIterator<Item = R>,
+{
```

Comment:
I think there are two choices:
1) As I did before use &str everywhere. This one is simple and IMHO sufficient.
2) Do as you say and make everything generic, but it has to be generic all the way down otherwise there is no reason to do it.(IMHO this change make code worse, but I want it to be merged ASAP)

Technical reason: even though trait bounds are the same types for this parameters should not be the same. It can be `&str`, `PathBuf`, `Path`, etc. e.g. mod_file_name can be `&str`, but input_dir can be `PathBuf`.

Reply:
@dvush I propose to finalize such signature like this: 
```rust
pub fn protobuf_generate<P, T, I>(input_dir: P, includes: I, mod_file_name: T)
 where
     P: AsRef<Path>,
     T: AsRef<str>,
     I: IntoIterator<Item = P>;
```
1. `input_dir` and `includes` are always paths and as for me they should bound to Path trait. Because path.  
2. `mode_file_name` - name of module. Should be `&str` or `String`.
@alekseysidorov @ilammy @pavel-mukhanov what do you think ?

Reply:
Yep. I've made a mistake.